File: javascript/src/router_bootstrap.ts
/**
 * FILE: javascript/src/router_bootstrap.ts
 * PURPOSE: Ensure the shared v5 router is reachable, or start it under the coordinated startup lock.
 * OWNS: Router port resolution, lock handling, log path resolution, readiness polling, and process bootstrap.
 * EXPORTS: resolveRouterPort, resolveRouterEndpoint, defaultRouterLockPath, defaultRouterLogPath, routerIsReachable, bootstrapRouter, RouterBootstrapResult
 * DOCS: docs/spec.md, rust/src/router_bootstrap.rs
 */

import { spawn, ChildProcess } from "child_process";
import { closeSync, openSync } from "fs";
import * as fs from "fs/promises";
import * as path from "path";
import * as os from "os";
import WebSocket from "ws";
import {
    DEFAULT_ROUTER_PORT,
    ROUTER_LOCK_SUFFIX,
    ROUTER_LOG_SUFFIX,
    ROUTER_PORT_ENV,
} from "./protocol.js";
import { BootstrapFailureError, BootstrapTimeoutError } from "./errors.js";

const DEFAULT_READY_TIMEOUT_MS = 10_000;
const DEFAULT_POLL_INTERVAL_MS = 100;
const DEFAULT_REACHABILITY_TIMEOUT_MS = 250;

export interface RouterBootstrapOptions {
    port?: number;
    routerBin?: string;
    lockPath?: string;
    logPath?: string;
    readinessTimeoutMs?: number;
    readinessPollIntervalMs?: number;
    reachabilityTimeoutMs?: number;
}

export interface RouterBootstrapResult {
    endpoint: string;
    port: number;
    lockPath: string;
    logPath: string;
    routerBin: string;
    started: boolean;
}

interface LockFileData {
    format: string;
    pid: number;
    router_pid: number | null;
    port: number;
    created_at_unix: number;
    expires_at_unix: number;
    status: string;
    language: string;
}

interface AcquiredLock {
    path: string;
    data: LockFileData;
    update(fields: Partial<LockFileData>): Promise<void>;
    release(): Promise<void>;
}

function sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
}

function toRouterPort(value: unknown, fallback: number): number {
    if (typeof value !== "string") {
        return fallback;
    }

    const parsed = Number.parseInt(value, 10);
    return Number.isFinite(parsed) && parsed > 0 ? parsed : fallback;
}

export function resolveRouterPort(options: RouterBootstrapOptions = {}): number {
    if (typeof options.port === "number" && Number.isFinite(options.port) && options.port > 0) {
        return options.port;
    }

    return toRouterPort(process.env[ROUTER_PORT_ENV], DEFAULT_ROUTER_PORT);
}

export function resolveRouterEndpoint(port: number = resolveRouterPort()): string {
    return `ws://127.0.0.1:${port}`;
}

export function defaultRouterLockPath(): string {
    return path.join(os.homedir(), ROUTER_LOCK_SUFFIX);
}

export function defaultRouterLogPath(): string {
    return path.join(os.homedir(), ROUTER_LOG_SUFFIX);
}

export function resolveRouterBin(options: RouterBootstrapOptions = {}): string {
    return options.routerBin ?? process.env.MULTIFROST_ROUTER_BIN ?? "multifrost-router";
}

export async function routerIsReachable(
    endpoint: string,
    timeoutMs: number = DEFAULT_REACHABILITY_TIMEOUT_MS
): Promise<boolean> {
    return await new Promise<boolean>(resolve => {
        const socket = new WebSocket(endpoint);
        let settled = false;

        const finish = (value: boolean) => {
            if (settled) return;
            settled = true;
            clearTimeout(timer);
            try {
                socket.terminate();
            } catch {
                // Ignore termination failures during probe cleanup.
            }
            resolve(value);
        };

        const timer = setTimeout(() => finish(false), timeoutMs);
        socket.once("open", () => finish(true));
        socket.once("error", () => finish(false));
        socket.once("close", () => finish(false));
    });
}

async function isProcessAlive(pid: number): Promise<boolean> {
    if (pid <= 0) return false;
    try {
        if (process.platform === "win32") {
            const { execFile } = await import("child_process");
            const { promisify } = await import("util");
            const execFileAsync = promisify(execFile);
            const { stdout } = await execFileAsync("tasklist", [
                "/FI",
                `PID eq ${pid}`,
                "/FO",
                "csv",
                "/NH",
            ]);
            return stdout.includes(`"${pid}"`);
        } else {
            return process.kill(pid, 0);
        }
    } catch {
        return false;
    }
}

async function evaluateExistingLock(lockPath: string): Promise<string> {
    let raw: string;
    try {
        raw = await fs.readFile(lockPath, "utf-8");
    } catch {
        return "reclaim";
    }

    let data: any;
    try {
        data = JSON.parse(raw);
    } catch {
        return "reclaim";
    }

    if (typeof data !== "object" || data.format !== "v1") {
        return "reclaim";
    }

    const now = Date.now() / 1000;
    if (data.expires_at_unix < now) {
        return "reclaim";
    }

    if (!(await isProcessAlive(data.pid))) {
        return "reclaim";
    }

    if (data.status === "failed") {
        return "reclaim";
    }

    if (data.router_pid != null && (await isProcessAlive(data.router_pid))) {
        return "skip_spawn";
    }

    return "wait";
}

async function acquireStartupLock(
    lockPath: string,
    port: number,
    timeoutMs: number
): Promise<AcquiredLock | null> {
    await fs.mkdir(path.dirname(lockPath), { recursive: true });

    const deadline = Date.now() + timeoutMs;

    while (true) {
        try {
            const lockData: LockFileData = {
                format: "v1",
                pid: process.pid,
                router_pid: null,
                port: port,
                created_at_unix: Date.now() / 1000,
                expires_at_unix: (Date.now() + timeoutMs) / 1000,
                status: "starting",
                language: "javascript",
            };

            // Atomic exclusive-create
            await fs.writeFile(lockPath, JSON.stringify(lockData, null, 2), { flag: "wx" });

            return {
                path: lockPath,
                data: lockData,
                update: async (fields: Partial<LockFileData>) => {
                    Object.assign(lockData, fields);
                    try {
                        await fs.writeFile(lockPath, JSON.stringify(lockData, null, 2));
                    } catch {
                        // best-effort
                    }
                },
                release: async () => {
                    try {
                        await fs.unlink(lockPath);
                    } catch {
                        // best-effort
                    }
                },
            };
        } catch (error: any) {
            if (error?.code !== "EEXIST" && error?.code !== "EACCES") {
                throw new BootstrapFailureError(`unexpected lock error: ${error.message}`, {
                    cause: error,
                });
            }

            // Evaluate existing lock
            const result = await evaluateExistingLock(lockPath);

            switch (result) {
                case "reclaim":
                    try {
                        await fs.unlink(lockPath);
                    } catch {
                        /* race, someone else got it */
                    }
                    continue;
                case "skip_spawn":
                    return null; // signal: don't spawn, wait for router
                case "wait":
                    if (Date.now() >= deadline) {
                        throw new BootstrapTimeoutError(
                            `timed out waiting for router bootstrap lock at ${lockPath}`
                        );
                    }
                    await sleep(DEFAULT_POLL_INTERVAL_MS);
                    continue;
            }
        }
    }
}

function spawnRouterProcess(routerBin: string, port: number, logPath: string): ChildProcess {
    const logFd = openSync(logPath, "a");
    const child = spawn(routerBin, [], {
        env: {
            ...process.env,
            [ROUTER_PORT_ENV]: String(port),
        },
        detached: true,
        stdio: ["ignore", logFd, logFd],
    });
    child.unref();

    child.once("exit", () => {
        try {
            closeSync(logFd);
        } catch {
            // Ignore close failures.
        }
    });

    child.once("error", () => {
        try {
            closeSync(logFd);
        } catch {
            // Ignore close failures.
        }
    });

    return child;
}

export async function bootstrapRouter(
    options: RouterBootstrapOptions = {}
): Promise<RouterBootstrapResult> {
    const port = resolveRouterPort(options);
    const endpoint = resolveRouterEndpoint(port);
    const lockPath = options.lockPath ?? defaultRouterLockPath();
    const logPath = options.logPath ?? defaultRouterLogPath();
    const routerBin = resolveRouterBin(options);
    const readinessTimeoutMs = options.readinessTimeoutMs ?? DEFAULT_READY_TIMEOUT_MS;
    const readinessPollIntervalMs = options.readinessPollIntervalMs ?? DEFAULT_POLL_INTERVAL_MS;
    const reachabilityTimeoutMs = options.reachabilityTimeoutMs ?? DEFAULT_REACHABILITY_TIMEOUT_MS;

    if (await routerIsReachable(endpoint, reachabilityTimeoutMs)) {
        return { endpoint, port, lockPath, logPath, routerBin, started: false };
    }

    const lock = await acquireStartupLock(lockPath, port, readinessTimeoutMs);

    // skip_spawn: someone else has an alive router, wait for it
    if (lock === null) {
        const deadline = Date.now() + readinessTimeoutMs;
        while (Date.now() < deadline) {
            if (await routerIsReachable(endpoint, reachabilityTimeoutMs)) {
                return { endpoint, port, lockPath, logPath, routerBin, started: false };
            }
            await sleep(readinessPollIntervalMs);
        }
        throw new BootstrapTimeoutError(
            `router did not become reachable at ${endpoint} within ${readinessTimeoutMs}ms`
        );
    }

    let child: ChildProcess | undefined;

    try {
        if (await routerIsReachable(endpoint, reachabilityTimeoutMs)) {
            await lock.update({ status: "ready" });
            return { endpoint, port, lockPath, logPath, routerBin, started: false };
        }

        child = spawnRouterProcess(routerBin, port, logPath);
        await lock.update({ router_pid: child.pid ?? null });

        const deadline = Date.now() + readinessTimeoutMs;
        let spawnError: Error | undefined;

        child.once("error", (error) => {
            spawnError = error instanceof Error ? error : new Error(String(error));
        });

        while (Date.now() < deadline) {
            if (spawnError) {
                await lock.update({ status: "failed" });
                throw new BootstrapFailureError(
                    `router process failed to start: ${spawnError.message}`,
                    { cause: spawnError }
                );
            }

            if (child.exitCode !== null) {
                await lock.update({ status: "failed" });
                throw new BootstrapFailureError(
                    `router process exited before readiness (code=${child.exitCode}, signal=${child.signalCode ?? "null"})`
                );
            }

            if (await routerIsReachable(endpoint, reachabilityTimeoutMs)) {
                await lock.update({ status: "ready" });
                return { endpoint, port, lockPath, logPath, routerBin, started: true };
            }

            await sleep(readinessPollIntervalMs);
        }

        // Timeout — router did not become reachable within the deadline
        await lock.update({ status: "failed" });
        try {
            child.kill();
        } catch {
            // Best effort only.
        }

        throw new BootstrapTimeoutError(
            `router did not become reachable at ${endpoint} within ${readinessTimeoutMs}ms`
        );
    } catch (error) {
        if (lock !== null) {
            await lock.update({ status: "failed" }).catch(() => {});
        }
        if (error instanceof BootstrapTimeoutError || error instanceof BootstrapFailureError) {
            throw error;
        }

        throw new BootstrapFailureError(
            error instanceof Error ? error.message : String(error),
            { cause: error }
        );
    } finally {
        await lock.release().catch(() => undefined);
    }
}

File: python/src/multifrost/router_bootstrap.py
"""
FILE: python/src/multifrost/router_bootstrap.py
PURPOSE: Discover, probe, and lazily start the shared v5 router process.
OWNS: Router endpoint resolution, startup lock handling, router reachability checks, and process spawning.
EXPORTS: RouterBootstrapConfig, router_port_from_env, default_router_lock_path, default_router_log_path, router_ws_url, router_is_reachable, bootstrap_router.
DOCS: docs/spec.md; agent_chat/python_v5_api_surface_2026-03-25.md
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import os
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from websockets.asyncio.client import connect as ws_connect
from websockets.exceptions import InvalidHandshake, InvalidURI, WebSocketException

from .errors import BootstrapError
from .protocol import (
    DEFAULT_ROUTER_PORT,
    ROUTER_BIN_ENV,
    ROUTER_LOCK_PATH_SUFFIX,
    ROUTER_LOG_PATH_SUFFIX,
    ROUTER_PORT_ENV,
)

_DEFAULT_READINESS_TIMEOUT = 10.0
_DEFAULT_POLL_INTERVAL = 0.1
_DEFAULT_LOCK_STALE_AFTER = 10.0
_ROUTER_HOST = "127.0.0.1"

# Fields present in every JSON lock file (for reference, not consumed by code).
LOCK_FILE_FIELDS = [
    "format",
    "pid",
    "router_pid",
    "port",
    "created_at_unix",
    "expires_at_unix",
    "status",
    "language",
]


def router_port_from_env() -> int:
    raw_value = os.environ.get(ROUTER_PORT_ENV)
    if raw_value is None or raw_value.strip() == "":
        return DEFAULT_ROUTER_PORT
    try:
        return int(raw_value)
    except ValueError:
        return DEFAULT_ROUTER_PORT


def default_router_lock_path() -> Path:
    return Path.home() / ROUTER_LOCK_PATH_SUFFIX


def default_router_log_path() -> Path:
    return Path.home() / ROUTER_LOG_PATH_SUFFIX


def router_ws_url(port: int | None = None) -> str:
    return f"ws://{_ROUTER_HOST}:{port or router_port_from_env()}"


def router_bin_path() -> str:
    override = os.environ.get(ROUTER_BIN_ENV)
    if override:
        return override
    return "multifrost-router"


@dataclass(slots=True)
class RouterBootstrapConfig:
    port: int = DEFAULT_ROUTER_PORT
    router_bin: str = field(default_factory=router_bin_path)
    lock_path: Path = field(default_factory=default_router_lock_path)
    log_path: Path = field(default_factory=default_router_log_path)
    readiness_timeout: float = _DEFAULT_READINESS_TIMEOUT
    poll_interval: float = _DEFAULT_POLL_INTERVAL
    lock_stale_after: float = _DEFAULT_LOCK_STALE_AFTER

    @property
    def endpoint(self) -> str:
        return router_ws_url(self.port)

    @classmethod
    def from_env(cls, port_override: int | None = None) -> RouterBootstrapConfig:
        port = port_override if port_override is not None else router_port_from_env()
        return cls(port=port)


async def router_is_reachable(endpoint: str) -> bool:
    try:
        async with ws_connect(
            endpoint,
            open_timeout=1.0,
            close_timeout=1.0,
            ping_interval=None,
            max_size=None,
        ):
            return True
    except (OSError, TimeoutError, InvalidHandshake, InvalidURI, WebSocketException):
        return False


async def bootstrap_router(
    config: RouterBootstrapConfig | None = None,
) -> RouterBootstrapConfig:
    config = config or RouterBootstrapConfig.from_env()
    if await router_is_reachable(config.endpoint):
        return config

    async with _StartupLock(
        config.lock_path,
        config.lock_stale_after,
        timeout=config.readiness_timeout,
        poll_interval=config.poll_interval,
        port=config.port,
    ) as lock:
        if lock._skip_spawn:
            deadline = time.monotonic() + config.readiness_timeout
            while time.monotonic() < deadline:
                if await router_is_reachable(config.endpoint):
                    return config
                await asyncio.sleep(config.poll_interval)
            raise BootstrapError(
                f"router did not become reachable at {config.endpoint} within "
                f"{config.readiness_timeout:.1f}s",
                timed_out=True,
            )

        if await router_is_reachable(config.endpoint):
            lock.update(status="ready")
            return config

        process = await _spawn_router_process(config)
        lock.update(router_pid=process.pid)

        try:
            await _wait_for_router_ready(config, process)
            lock.update(status="ready")
        except Exception:
            lock.update(status="failed")
            await _terminate_router_process(process)
            raise

    return config


async def _wait_for_router_ready(
    config: RouterBootstrapConfig, process: asyncio.subprocess.Process
) -> None:
    deadline = time.monotonic() + config.readiness_timeout
    while time.monotonic() < deadline:
        if await router_is_reachable(config.endpoint):
            return
        if process.returncode is not None:
            raise BootstrapError(
                f"router exited before becoming reachable (code={process.returncode})"
            )
        await asyncio.sleep(config.poll_interval)

    raise BootstrapError(
        f"router did not become reachable at {config.endpoint} within "
        f"{config.readiness_timeout:.1f}s",
        timed_out=True,
    )


async def _spawn_router_process(
    config: RouterBootstrapConfig,
) -> asyncio.subprocess.Process:
    config.lock_path.parent.mkdir(parents=True, exist_ok=True)
    config.log_path.parent.mkdir(parents=True, exist_ok=True)
    log_file = config.log_path.open("ab", buffering=0)
    env = os.environ.copy()
    env[ROUTER_PORT_ENV] = str(config.port)

    try:
        process = await asyncio.create_subprocess_exec(
            config.router_bin,
            stdout=log_file,
            stderr=log_file,
            env=env,
        )
    except FileNotFoundError as err:
        raise BootstrapError(
            f"router binary not found: {config.router_bin}", cause=err
        ) from err
    except Exception as err:  # pragma: no cover - defensive
        raise BootstrapError(f"failed to start router: {err}", cause=err) from err
    finally:
        log_file.close()

    return process


async def _terminate_router_process(process: asyncio.subprocess.Process) -> None:
    if process.returncode is not None:
        return
    process.kill()
    with contextlib.suppress(ProcessLookupError):
        await process.wait()


@dataclass(slots=True)
class _StartupLock:
    path: Path
    stale_after: float
    timeout: float = _DEFAULT_READINESS_TIMEOUT
    poll_interval: float = _DEFAULT_POLL_INTERVAL
    port: int = DEFAULT_ROUTER_PORT
    _fd: int | None = field(init=False, default=None)
    _skip_spawn: bool = field(init=False, default=False)

    async def __aenter__(self) -> _StartupLock:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        deadline = time.monotonic() + self.timeout
        while True:
            try:
                self._fd = os.open(
                    self.path,
                    os.O_CREAT | os.O_EXCL | os.O_WRONLY,
                    0o644,
                )
            except FileExistsError:
                if time.monotonic() >= deadline:
                    raise BootstrapError(
                        "timed out waiting for router bootstrap lock"
                    ) from None

                match self._evaluate_existing_lock():
                    case "reclaim":
                        with contextlib.suppress(FileNotFoundError):
                            self.path.unlink()
                        continue
                    case "skip_spawn":
                        self._skip_spawn = True
                        return self
                    case "wait":
                        await asyncio.sleep(self.poll_interval)
                        continue
                    case _:
                        await asyncio.sleep(self.poll_interval)
                        continue

            lock_data = {
                "format": "v1",
                "pid": os.getpid(),
                "router_pid": None,
                "port": self.port,
                "created_at_unix": time.time(),
                "expires_at_unix": time.time() + self.timeout,
                "status": "starting",
                "language": "python",
            }
            os.write(self._fd, json.dumps(lock_data).encode())
            os.close(self._fd)
            self._fd = None
            return self

    async def __aexit__(self, exc_type: Any, exc: Any, tb: Any) -> None:
        if not self._skip_spawn:
            if exc_type is not None:
                self.update(status="failed")
            self.release()

    def _evaluate_existing_lock(self) -> str:
        """Apply the 6 stale-detection rules from the Lock File Format spec.

        Returns one of:
        - "reclaim"    — lock is stale/expired/we can take it
        - "skip_spawn" — holder has a live router, try connecting first
        - "wait"       — lock is valid, holder is alive, retry later
        """
        try:
            raw = self.path.read_text()
            data = json.loads(raw)
        except (OSError, json.JSONDecodeError, ValueError):
            return "reclaim"

        if not isinstance(data, dict) or data.get("format") != "v1":
            return "reclaim"

        if data.get("expires_at_unix", 0) < time.time():
            return "reclaim"

        if not self._is_process_alive(data.get("pid", 0)):
            return "reclaim"

        if data.get("status") == "failed":
            return "reclaim"

        router_pid = data.get("router_pid")
        if router_pid is not None and self._is_process_alive(router_pid):
            return "skip_spawn"

        return "wait"

    @staticmethod
    def _is_process_alive(pid: int) -> bool:
        if pid <= 0:
            return False
        if os.name == "nt":  # Windows
            try:
                import subprocess

                result = subprocess.run(
                    ["tasklist", "/FI", f"PID eq {pid}", "/FO", "csv", "/NH"],
                    capture_output=True,
                    text=True,
                    timeout=5,
                )
                return str(pid) in result.stdout
            except Exception:
                return False
        else:  # Unix
            try:
                os.kill(pid, 0)
                return True
            except OSError:
                return False

    def update(self, **fields: Any) -> None:
        """Update lock file fields in-place (best-effort, no throw)."""
        try:
            raw = self.path.read_text()
            data = json.loads(raw)
        except Exception:
            return
        data.update(fields)
        with contextlib.suppress(Exception):
            self.path.write_text(json.dumps(data))

    def release(self) -> None:
        if self._fd is not None:
            with contextlib.suppress(OSError):
                os.close(self._fd)
            self._fd = None
        with contextlib.suppress(FileNotFoundError):
            self.path.unlink()


__all__ = [
    "RouterBootstrapConfig",
    "router_port_from_env",
    "default_router_lock_path",
    "default_router_log_path",
    "router_ws_url",
    "router_bin_path",
    "router_is_reachable",
    "bootstrap_router",
]
